lib/flows/cron/data_retention.py

```diff
@@ -66,3 +65,65 @@ def Start(self):
     for obj in jobs_objs:
       age = rdfvalue.RDFDatetime().Now() - cron_jobs_ttl
       obj.DeleteJobFlows(age)
+      self.HeartBeat
+
+class CleanTemp(cronjobs.SystemCronFlow):
+  """Cleaner that deletes temp objects."""
+
+  frequency = rdfvalue.Duration("7d")
+  lifetime = rdfvalue.Duration("1d")
+
+  @flow.StateHandler()
+  def Start(self):
+    tmp_ttl = config_lib.CONFIG["DataRetention.tmp_ttl"]
+    if not tmp_ttl:
+      self.Log("TTL not set - nothing to do...")
+      return
+
+    exception_label = config_lib.CONFIG[
+        "DataRetention.tmp_ttl_exception_label"]
+
+    tmp_root = aff4.FACTORY.Open("aff4:/tmp", mode="r", token=self.token)
+    tmp_urns = list(tmp_root.ListChildren())
+
+    deadline = rdfvalue.RDFDatetime().Now() - tmp_ttl
+
+    for urn in tmp_urns:
+      obj = aff4.FACTORY.Open(urn, token=self.token)
+      if exception_label in obj.GetLabelsNames():
+        continue
+
+      if urn.age < deadline:
+        aff4.FACTORY.Delete(urn, token=self.token)
+        self.HeartBeat()
+
+class CleanInactiveClients(cronjobs.SystemCronFlow):
+  """Cleaner that deletes inactive clients."""
+
+  frequency = rdfvalue.Duration("7d")
+  lifetime = rdfvalue.Duration("1d")
+
+  @flow.StateHandler()
+  def Start(self):
+    inactive_client_ttl = config_lib.CONFIG["DataRetention.inactive_client_ttl"]
+    if not inactive_client_ttl:
+      self.Log("TTL not set - nothing to do...")
+      return
+
+    exception_label = config_lib.CONFIG[
+        "DataRetention.inactive_client_ttl_exception_label"]
+
+    index = aff4.FACTORY.Open(client_index.MAIN_INDEX, mode="r",
+                               token=self.token)
+    client_urns = index.LookupClients(["."])
+
+    deadline = rdfvalue.RDFDatetime().Now() - inactive_client_ttl
+
+    for client_urn in client_urns:
+      client = aff4.FACTORY.Open(client_urn, mode="r", token=self.token)
```

Comment:
Opening clients one by one will not scale, please group them in batches and open using MultiOpen.
Example code that does this is here:
https://github.com/google/grr/blob/master/lib/hunts/implementation.py#L982

You can use a much bigger chunk size than we use there, maybe 10000 would be a good number.

Once you have all clients open, please collect the ones to delete and use aff4.FACTORY.MultiDelete. Should also be much faster.